lib/polygeist/Ops.cpp

```diff
@@ -194,6 +194,116 @@ class SubToCast final : public OpRewritePattern<SubIndexOp> {
+    auto dims = srcMemRefType.getShape().size();
+
+    // For now, restrict subview lowering to statically defined memref's
+    if (!srcMemRefType.hasStaticShape() | !resMemRefType.hasStaticShape())
+      return failure();
+
+    // For now, restrict to simple rank-reducing indexing
+    if (srcMemRefType.getShape().size() <= resMemRefType.getShape().size())
+      return failure();
+
+    // Build offset, sizes and strides
+    SmallVector<OpFoldResult> sizes(dims, rewriter.getIndexAttr(0));
+    sizes[0] = op.index();
+    SmallVector<OpFoldResult> offsets(dims);
+    for (auto dim : llvm::enumerate(srcMemRefType.getShape())) {
+      if (dim.index() == 0)
+        offsets[0] = rewriter.getIndexAttr(1);
+      else
+        offsets[dim.index()] = rewriter.getIndexAttr(dim.value());
+    }
+    SmallVector<OpFoldResult> strides(dims, rewriter.getIndexAttr(1));
+
+    // Generate the appropriate return type:
+    auto subMemRefType = MemRefType::get(srcMemRefType.getShape().drop_front(),
+                                         srcMemRefType.getElementType());
+
+    rewriter.replaceOpWithNewOp<memref::SubViewOp>(
+        op, subMemRefType, op.source(), sizes, offsets, strides);
+
+    return success();
+  }
+};
+
+// Simplify redundant dynamic subindex patterns which tries to represent
+// rank-reducing indexing:
+//   %3 = "polygeist.subindex"(%1, %arg0) : (memref<2x1000xi32>, index) ->
+//   memref<?x1000xi32> %4 = "polygeist.subindex"(%3, %c0) :
+//   (memref<?x1000xi32>, index) -> memref<1000xi32>
+// simplifies to:
+//   %4 = "polygeist.subindex"(%1, %arg0) : (memref<2x1000xi32>, index) ->
+//   memref<1000xi32>
+
+class RedundantDynSubIndex final : public OpRewritePattern<SubIndexOp> {
+public:
+  using OpRewritePattern<SubIndexOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(SubIndexOp op,
+                                PatternRewriter &rewriter) const override {
+    auto srcOp = dyn_cast<SubIndexOp>(op.source().getDefiningOp());
+    if (!srcOp)
+      return failure();
+
+    auto srcMemRefType = op.source().getType().cast<MemRefType>();
+    auto resMemRefType = op.result().getType().cast<MemRefType>();
+
+    // Check if there are multiple users of the dynamically sized memory
+    if (!op.source().hasOneUse())
+      return failure();
+
+    // Check that the source op indeed is a dynamically indexed memory in the
+    // 0'th index.
+    if (srcMemRefType.getShape()[0] != -1)
+      return failure();
+
+    // Check that this is indeed a rank reducing operation
+    if (srcMemRefType.getShape().size() !=
+        (resMemRefType.getShape().size() + 1))
+      return failure();
+
+    // Check that there is not a downstream cast of subindex result. This is a
+    // bit dubious, but allowing cast canonicalizations - when possible - to
+    // convert subindexes will ultimately result in fewer memref.subview
+    // operations to be inferred.
+    for (auto user : op.getResult().getUsers()) {
```

Comment:
Agreed; quite odd. It's due to (as i see it) unintended mixing of the various canonicalization patterns for the SubIndexOp + how subindex lowering is done internally in Polygeist. 

An input could be:
```c
void matrix_power(int x[20][20], int row[20], int col[20], int a[20]) {
  for (int k = 1; k < 20; k++) {
    for (int p = 0; p < 20; p++) {
      x[k][row[p]] += a[p] * x[k - 1][col[p]];
    }
  }
}
```

**With** this check, we get:
```mlir
  func @matrix_power(%arg0: memref<20x20xi32>, %arg1: memref<20xi32>, %arg2: memref<20xi32>, %arg3: memref<20xi32>) attributes {llvm.linkage = #llvm.linkage<external>} {
    %c20 = arith.constant 20 : index
    %c0 = arith.constant 0 : index
    %c1 = arith.constant 1 : index
    %c1_i32 = arith.constant 1 : i32
    scf.for %arg4 = %c1 to %c20 step %c1 {
      %0 = arith.subi %arg4, %c1 : index
      %1 = arith.index_cast %0 : index to i32
      %2 = arith.addi %1, %c1_i32 : i32
      %3 = arith.index_cast %2 : i32 to index
      scf.for %arg5 = %c0 to %c20 step %c1 {
        %4 = memref.load %arg1[%arg5] : memref<20xi32>
        %5 = arith.index_cast %4 : i32 to index
        %6 = memref.load %arg3[%arg5] : memref<20xi32>
        %7 = arith.subi %3, %c1 : index
        %8 = memref.load %arg2[%arg5] : memref<20xi32>
        %9 = arith.index_cast %8 : i32 to index
        %10 = memref.load %arg0[%7, %9] : memref<20x20xi32>
        %11 = arith.muli %6, %10 : i32
        %12 = memref.load %arg0[%3, %5] : memref<20x20xi32>
        %13 = arith.addi %12, %11 : i32
        memref.store %13, %arg0[%3, %5] : memref<20x20xi32>
      }
    }
    return
  }
```

However, when we omit the check we get:
```mlir
  func @matrix_power(%arg0: memref<20x20xi32>, %arg1: memref<20xi32>, %arg2: memref<20xi32>, %arg3: memref<20xi32>) attributes {llvm.linkage = #llvm.linkage<external>} {
    %c20 = arith.constant 20 : index
    %c0 = arith.constant 0 : index
    %c1 = arith.constant 1 : index
    %c1_i32 = arith.constant 1 : i32
    scf.for %arg4 = %c1 to %c20 step %c1 {
      %0 = arith.subi %arg4, %c1 : index
      %1 = arith.index_cast %0 : index to i32
      %2 = arith.addi %1, %c1_i32 : i32
      %3 = arith.index_cast %2 : i32 to index
      scf.for %arg5 = %c0 to %c20 step %c1 {
        %4 = memref.subview %arg0[%3, 0] [1, 20] [1, 1] : memref<20x20xi32> to memref<20xi32> // <- unintentional
        %5 = memref.load %arg1[%arg5] : memref<20xi32>
        %6 = arith.index_cast %5 : i32 to index
        %7 = memref.load %arg3[%arg5] : memref<20xi32>
        %8 = arith.subi %3, %c1 : index
        %9 = memref.subview %arg0[%8, 0] [1, 20] [1, 1] : memref<20x20xi32> to memref<20xi32>
        %10 = memref.load %arg2[%arg5] : memref<20xi32>
        %11 = arith.index_cast %10 : i32 to index
        %12 = memref.load %9[%11] : memref<20xi32>
        %13 = arith.muli %7, %12 : i32
        %14 = memref.load %4[%6] : memref<20xi32> // <- unintentional
        %15 = arith.addi %14, %13 : i32
        memref.store %15, %4[%6] : memref<20xi32> // <- unintentional
      }
    }
    return
  }
```
I see this as an unintentional result - i.e. a subview is used and the load/stores uses the subview instead of a multidimensional memref. I'd ideally like to see the output program use as high-level memory semantics as possible. We can always flatten the memories and indexing into these afterwards, but regenerating that information is hard.

This seems to happen because, during canonicalization, a part of the program looks like this:
```mlir
...
  %56 = "polygeist.subindex"(%52, %55) : (memref<20x20xi32>, index) -> memref<?x20xi32>
  %57 = "polygeist.subindex"(%56, %c0) : (memref<?x20xi32>, index) -> memref<20xi32>
  %58 = memref.cast %57 : memref<20xi32> to memref<?xi32>
  %59 = memref.load %8[%c0] : memref<?xmemref<20xi32>>
...
```
which will trigger the canonicalization pattern and insert a `memref.subview`. However, we'd like to have `CastOfSubIndex` apply here, which will remove the interim subindex.
I tried getting around this by adjusting the benefit values for the different patterns such as the cast(subindex(...), but that unfortunately didn't work.

Reply:
I see. So essentially this is done to force the subindex/cast to bind more strongly (and force the store/load subindex simplification to run).

This makes sense then (unless there's other ways of setting priorities). Alternatively, we can enable a subview/load simplification similar to the one we have for subindex.